tests/actions/SessionTest.js

```diff
@@ -23,81 +22,77 @@ Onyx.init({
+            .then(() => {
+                // Then our re-authentication credentials should be generated and our session data
+                // have the correct information + initial authToken.
+                expect(credentials.login).toBe(TEST_USER_LOGIN);
+                expect(credentials.autoGeneratedLogin).not.toBeUndefined();
+                expect(credentials.autoGeneratedPassword).not.toBeUndefined();
+                expect(session.authToken).toBe(TEST_INITIAL_AUTH_TOKEN);
+                expect(session.accountID).toBe(TEST_USER_ACCOUNT_ID);
+                expect(session.email).toBe(TEST_USER_LOGIN);
+
+                // At this point we have an authToken. To simulate it expiring we'll just make another
+                // request and mock the response so it returns 407. Once this happens we should attempt
+                // to Re-Authenticate with the stored credentials. Our next call will be to Authenticate
+                // so we will mock that response with a new authToken and then verify that Onyx has our
+                // data.
+                HttpUtils.xhr
+
+                    // This will make the call to DeprecatedAPI.Get() below return with an expired session code
+                    .mockImplementationOnce(() => Promise.resolve({
+                        jsonCode: CONST.JSON_CODE.NOT_AUTHENTICATED,
+                    }))
+
+                    // The next call should be Authenticate since we are reauthenticating
+                    .mockImplementationOnce(() => Promise.resolve({
+                        jsonCode: CONST.JSON_CODE.SUCCESS,
+                        accountID: TEST_USER_ACCOUNT_ID,
+                        authToken: TEST_REFRESHED_AUTH_TOKEN,
+                        email: TEST_USER_LOGIN,
+                    }));
+
+                // When we attempt to fetch the chatList via the API
+                DeprecatedAPI.Get({returnValueList: 'chatList'});
+                return waitForPromisesToResolve();
+            })
+            .then(() => {
+                // Then it should fail and reauthenticate the user adding the new authToken to the session
+                // data in Onyx
+                expect(session.authToken).toBe(TEST_REFRESHED_AUTH_TOKEN);
+            });
     });
-});
 
-test('Push notifications are unsubscribed after signing out', () => {
-    TestHelper.signInWithTestUser()
-        .then(() => Session.signOut());
+    test('Push notifications are subscribed after signing in', () => (
+        TestHelper.signInWithTestUser()
+            .then(() => expect(PushNotification.register).toBeCalled())
+    ));
 
-    return waitForPromisesToResolve().then(() => {
-        expect(PushNotification.deregister).toBeCalled();
-    });
+    test('Push notifications are unsubscribed after signing out', () => (
+        TestHelper.signInWithTestUser()
+            .then(TestHelper.signOutTestUser)
```

Comment:
Why we didn't use `HttpUtils.xhr.mockImplementationOnce` and then just call `Session.signOutAndRedirectToSignIn()`? is it because we do more than one request during this process?

Reply:
Few reasons:

- I think that this is more readable and likely has better re-use value.
- `TestHelper.signOutTestUser` contains some additional logic for waiting for promises to resolve
- `TestHelp.signOutTestUser` is self-contained in that it restores HttpsUtils.xhr back to it's completely un-mocked state when it's done.

Reply:
Sounds good, thanks